crates/solver/src/settlement_post_processing/optimize_unwrapping.rs

```diff
@@ -0,0 +1,206 @@
+use crate::settlement::Settlement;
+use anyhow::Result;
+use contracts::WETH9;
+use primitive_types::U256;
+
+/// Tries to do one of 2 optimizations.
+/// 1) Drop WETH unwraps and instead pay ETH with the settlment contract's buffer.
+/// 2) Top up settlement contract's ETH buffer by unwrapping way more WETH than this settlement
+///    needs. This will cause the next few settlements to use optimization 1.
+/// If this function returns Err(), the original settlement will not be modified in any way.
+pub async fn optimize_unwrapping<V, VFut, B, BFut>(
+    settlement: &mut Settlement,
+    settlement_would_succeed: V,
+    get_weth_balance: B,
```

Comment:
Cool use of closures to make the method testable instead of using a trait and mocks.

Reply:
> It might be easier to extract the IO you need into a trait that can be mocked. You can look at other examples in the code where use mockall::automock.

Personally, I'm OK either way. I would only slightly prefer mocking just to be more consistent with the rest of the codebase.

Reply:
I refactored the code to use traits and automock. Although it was possible to test the code with the closures, it made the signature of the optimization function very complex. Also having an automocked trait return different values across multiple invocations is way nicer than make a closure do that.